### PR TITLE
Party creation

### PR DESCRIPTION
**Pull request by dwatchguard**
Basic Party Functionality has been added.

Map has a container of Party.
Party has a List of Unit

Party can Iterate through Units, take Unit from another Party, and can create Unit from scratch.

Units can now move through allied Units.

### CODE REVIEW

File: Project1/CellEdge.cpp [hunk 1]
```diff
@@ -1,24 +1,39 @@
...
 
-std::optional<int> CellEdge::getCost(MobilityList<bool> mobility_type) const {
-	return getCost(mobility_type, true);
+std::optional<int> CellEdge::getCost(Mobility mobility)
+{
+	std::optional<int> cost;
+	for (MobilityType mobility_type : MobilityType::list) {
+		if (mobility.canPass(mobility_type, _cell.getTile()._unit)) {
+			std::optional<int> edge_cost = getCost(mobility_type);
+			if (edge_cost.has_value() && cost < edge_cost) { // if we can step on the tile
```
Comment (SilensAngelusNex) [inline Project1/CellEdge.cpp]: Doesn't this mean you find the _greatest_ edge cost?

Should be `edge_cost.has_value() && (cost > edge_cost || !cost.has_value())`?

File: Project1/CellEdge.cpp [hunk 2]
```diff
@@ -1,24 +1,39 @@
...
 
-std::optional<int> CellEdge::getCost(MobilityType mobility) const {
-	return _costs[mobility];
+std::optional<int> CellEdge::getCost(MobilityType mobility_type) const {
+	return _costs[mobility_type];
 }
 
-std::optional<int> CellEdge::getCost(MobilityList<bool> mobility_type) const {
-	return getCost(mobility_type, true);
+std::optional<int> CellEdge::getCost(Mobility mobility)
```
Comment (SilensAngelusNex) [inline Project1/CellEdge.cpp]: I think you should move most of this method into Mobility.

`Mobility::getCost(CellEdge)` or something.

File: Project1/CellEdge.h [hunk 3]
```diff
@@ -1,11 +1,10 @@
...
-#include <array>
 #include "MobilityList.h"
+#include "Mobility.h"
 #include <optional>
-/**
-Represents an Edge to a GridCell, basically a struct
-*/
+
+class GridCell;
+//typedef bool(*boolFunc)(Unit* param);
```
Comment (SilensAngelusNex) [inline Project1/CellEdge.h]: What was this for?

File: Project1/Map.h [hunk 4]
```diff
@@ -18,36 +23,45 @@ class Map {
...
-	CellPath getShortestPath(GridCell & start, GridCell & destination, int max_move, MobilityList<bool> mobility);
+	CellPath getShortestPath(GridCell & start, GridCell & destination, int max_move, MobilitySet mobility);
 	
-	//I can't get std::as_const to work
 	const PathMap findShortestPaths(GridCell& start) const;
-	const PathMap findShortestPaths(GridCell& start, int max_move, MobilityList<bool> mobility) const;
-	std::vector<GridCell*> getAlliedCells(GridCell& unit_cell);
+	const PathMap findShortestPaths(GridCell& start, int max_move, MobilitySet mobility) const;
+	const PathMap findShortestPaths(GridCell& start, Mobility& mobility) const;
+	const PathMap findShortestPaths(GridCell & start, int max_move, MobilitySet mobility, bool intangible) const;
```
Comment (SilensAngelusNex) [inline Project1/Map.h:63]: Should these not all be `const GridCell&`? Did we talk about this?

Edit: Won't it work if you make `PathMap = map<const GridCell*, CellEdge>`?

File: Project1/CellEdge.cpp [hunk 5]
```diff
@@ -1,21 +1,33 @@
...
 	_costs(costs)
 {}
 
-std::optional<int> CellEdge::getCost(MobilityType mobility) const {
+std::optional<int> CellEdge::getCost(const MobilityType mobility) const {
 	return _costs[mobility];
 }
-
-std::optional<int> CellEdge::getCost(MobilityList<bool> mobility_type) const {
+std::optional<int> CellEdge::getCost(const Mobility mobility) const {
```
Comment (SilensAngelusNex) [inline Project1/CellEdge.cpp]: I think some/all of this method should move to Mobility.
`Mobility::getCost(CellEdge)`

File: Project1/CellEdge.h [hunk 6]
```diff
@@ -16,9 +17,10 @@ class CellEdge {
 
 	GridCell& _cell;
 	CellEdge(GridCell& cell, MobilityList<std::optional<int>> costs);
-	std::optional<int> getCost(MobilityType mobility) const;
-	std::optional<int> getCost(MobilityList<bool> mobility_type) const;
-	std::optional<int> getCost(MobilityList<bool> mobility_type, bool intangible) const;
+	std::optional<int> getCost(const MobilityType mobility_type) const;
+	std::optional<int> getCost(const Mobility mobility) const;
+	std::optional<int> getCost(const MobilitySet mobility_type) const;
+	std::optional<int> getCost(const MobilitySet mobility_type, bool intangible) const;
```
Comment (SilensAngelusNex) [inline Project1/CellEdge.h]: _Either_ take the argument by value _or_ make it const. If you're taking by value, you're copying anyway, so you can't modify what the caller passed in.

MobilityType should def be by value, MobilitySet too probably, because they're the same size as an pointer(and therefore reference) so you don't save anything passing by reference.
Mobility should be const ref because you can't copy them.

File: Project1/Map.cpp [hunk 7]
```diff
@@ -48,30 +66,49 @@ GridCell& Map::getGridCell(int x_pos, int y_pos) {
 	return _grid[x_pos][y_pos];
 }
 GridCell& Map::getGridCell(Unit & unit) {
+	Expects(hasUnit(unit));
 	return *_unit_to_cell[&unit];
 }
-
+bool Map::hasUnit(Unit& unit) const{
+	return _unit_to_cell.count(&unit) == 1;
```
Comment (SilensAngelusNex) [inline Project1/Map.cpp]: `> 0`

super confusing if we ever mess up and put the same one in twice somehow

File: Project1/Map.cpp [hunk 8]
```diff
@@ -48,30 +66,49 @@ GridCell& Map::getGridCell(int x_pos, int y_pos) {
...
 const GridCell& Map::getGridCell(Unit & unit) const{
 	return *_unit_to_cell.at(&unit);
 }
+const Party& Map::getParty(Unit & unit) const {
+	return unit._party->getParty(Passkey<Map>()); //This is probably bad
+}
 ///////////////////////////////////////////////////////////
 PathMap Map::findShortestPaths(GridCell& start) {
-	return findShortestPaths(start, INT_MAX, MobilityList<bool>({true, false, false}));
+	return std::as_const(*this).findShortestPaths(start);
```
Comment (SilensAngelusNex) [inline Project1/Map.cpp:90]: What does this do exactly? Why can't you just call `this->findShortestPaths(start)`?

File: Project1/Mobility.cpp [hunk 9]
```diff
@@ -1,31 +1,33 @@
...
 }
 
-int& Mobility::getMove()
-{
+int& Mobility::getMove() {
 	return _movement;
+}
+
+bool Mobility::canPass(MobilityType mobility_type, const Unit* unit) const {
+	return _mobility[mobility_type] && (unit == nullptr ||_owner._party->hasUnit(*unit));
```
Comment (SilensAngelusNex) [inline Project1/Mobility.cpp]: This is going to need an additional condition OR'd on in the future for skills. Can you mark it?

Also, if you move the getCost function into here you won't need to check each MobilityType separately.

File: Project1/Mobility.h [hunk 10]
```diff
@@ -1,18 +1,25 @@
 #pragma once
 #include "MobilityList.h"
-class Mobility
-{
+#include "Component.h"
+
+class Unit;
+
+class Mobility : Component<Unit> {
```
Comment (SilensAngelusNex) [inline Project1/Mobility.h]: Did you use `_owner` at all? What was the rationale behind making this a Component?

File: Project1/MoveHelper.h [hunk 11]
```diff
@@ -1,8 +1,12 @@
...
 #include "Map.h"
-#include "CellPath.h"
-class MoveHelper
-{
+
+class Map;
+class CellPath;
+class GridCell;
+class Unit;
+class MoveHelper {
```
Comment (SilensAngelusNex) [inline Project1/MoveHelper.h:9]: Why isn't this a map subclass?

Why are you both including _and_ forward declaring Map?

File: Project1/MoveHelper.cpp [hunk 12]
```diff
@@ -68,7 +72,20 @@ std::vector<GridCell*> MoveHelper::getAllAttackableCells(Unit& unit) {
...
+	auto vec = std::vector<GridCell*>();
+	for (Unit& ally : _map.getParty(unit).getUnits()) {
+		vec.push_back(&_map.getGridCell(ally));
+	}
+	return vec;
+}
+std::vector<GridCell*> MoveHelper::getOtherAlliedCells(Unit& unit) {
+	auto vec = std::vector<GridCell*>();
+	for (Unit& ally : _map.getParty(unit).getOtherUnits(unit)) {
+		if (_map.hasUnit(ally)) {
```
Comment (SilensAngelusNex) [inline Project1/MoveHelper.cpp:86]: Why do you not need this line in getAlliedCells?

File: Project1/PartyBase.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	_party_name(std::move(name)) {
+	for (UnitData& data : unit_data) {
+		insertUnit(data);
+	}
+}
+PartyBase::PartyBase(PartyData data) :
+	PartyBase(data.name, data.unit_data)
+{}
+
+PartyBase::const_iterator PartyBase::getPosition(const Unit & unit) const {
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.cpp:21]: What are these for? You give it a Unit and they return a iterator that points to that unit?

File: Project1/PartyBase.h [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+	PartyBase();
+	PartyBase(std::string name);
+	PartyBase(std::string name, std::vector<UnitData> unit_data);
+	PartyBase(PartyData data);
+
+	const_iterator getPosition(const Unit& unit) const;
+	iterator getPosition(const Unit& unit);
+
+	void insertUnit(UnitData unit);
+	void insertUnit(Unit& unit);
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.h]: This should take `unique_ptr<Unit>`, because we want the Party to own it's units instead of just knowing about units someone else owns.

File: Project1/Terrain.h [hunk 15]
```diff
@@ -2,12 +2,11 @@
...
-//TODO(Torrey): FACTORY
-enum class _TERRAIN_TYPES { PLAINS, WATER, FOREST, MOUNTAIN, WALL, ENUM_END }; //Temporary Solution
+
+enum class _TERRAIN_TYPES { PLAINS, WATER, FOREST, MOUNTAIN, WALL, ENUM_END }; 
 using TerrainType = Enum<_TERRAIN_TYPES>;
 class Terrain {
 private:
-	using empty = std::optional<int>; //static const std::optional<int> empty = std::optional<int>();??
 	std::string _name;
 	MobilityList<std::optional<int>> _costs;
```
Comment (SilensAngelusNex) [inline Project1/Terrain.h]: Is MobilityList<optional<int>> used enough that it deserves it's own name like MobilitySet?

File: Project1/Unit.cpp [hunk 16]
```diff
@@ -45,6 +45,15 @@ Combat& Unit::getCombat() {
...
 }
 
+void Unit::refresh() {
+	//TODO(T or W): Code start of turn stuff.
+}
+void Unit::newTurn() {
+	//TODO(T or W): code start of turn stuff.
+}
+bool Unit::isTired() const {
+	return true; //TODO(Weston): Should this be in a component like isDead() ?
```
Comment (SilensAngelusNex) [inline Project1/Unit.cpp:55]: Not in with isDead, but we should put all the level/turn info together.
Has attacked? and Movement remaining are the ones I can think of off the top of my head, but I vaguely remember wanting other ones too.

File: Project1/Unit.cpp [hunk 17]
```diff
@@ -68,3 +77,12 @@ Inventory& Unit::getInventoryInternal() {
 CONTEXT& Unit::getContext() {
 	return _context;
 }
+
+bool Unit::operator==(const Unit& unit) const {
+	return this == &unit;
```
Comment (SilensAngelusNex) [inline Project1/Unit.cpp:82]: Hello, Java, my old friend....

File: Project1/Unit.h [hunk 18]
```diff
@@ -14,10 +14,16 @@
 #include "Experience.h"
 #include "Mobility.h"
 #include "Health.h"
+#include "PartyBase.h"
 
+class PartyBase;
 using IDENTITY = std::string;
 using CONTEXT = Dice<100>;
-
+struct UnitData {
```
Comment (SilensAngelusNex) [inline Project1/Unit.h:22]: Why? Just to make construction easier?

File: Project1/CellPath.h [hunk 19]
```diff
@@ -1,17 +1,19 @@
 #pragma once
-#include "GridCell.h"
+#include <list>
+#include "MobilityList.h"
+class GridCell;
 //TODO(Torrey): Create Superclasses of this. EX: UnitPath has a unit to traverse the Map.
 using CellWrap = std::reference_wrapper<GridCell>;
```
Comment (SilensAngelusNex) [inline Project1/CellPath.h]: Can we rename this CellRef? I think I saw somewhere else where you used that naming convention for a reference wrapper and I like it better.

**Review by SilensAngelusNex — COMMENTED** [review]
Aight, think I'm done

File: Project1/Map.h (hunk 4 above)
Comment (dwatchguard) [reply Project1/Map.h:63]: prolly

File: Project1/Mobility.h (hunk 10 above)
Comment (dwatchguard) [reply Project1/Mobility.h]: canPass(MobilityType, const Unit*) uses it

File: Project1/MoveHelper.cpp (hunk 12 above)
Comment (dwatchguard) [reply Project1/MoveHelper.cpp:86]: good questions

File: Project1/MoveHelper.h (hunk 11 above)
Comment (dwatchguard) [reply Project1/MoveHelper.h:9]: I am refactoring this in a new branch.

It isn't a subclass because i want to have different helper interfaces for movement, combat, and skills. and one map cannot branch into multiple inheritance directions.

I am making a MapHelper class that the other interfaces will inherit.

File: Project1/Terrain.h (hunk 15 above)
Comment (dwatchguard) [reply Project1/Terrain.h]: Probably, I'll add that in

File: Project1/Mobility.h (hunk 10 above)
Comment (SilensAngelusNex) [reply Project1/Mobility.h]: Gotcha

File: Project1/Unit.h (hunk 18 above)
Comment (dwatchguard) [reply Project1/Unit.h:22]: Yeah.
It also makes it easier to construct a bunch from a file (like when initializing a party )

File: Project1/Map.cpp (hunk 8 above)
Comment (dwatchguard) [reply Project1/Map.cpp:90]: this->findShortestPaths(start) would be a recursive nightmare is why. as_const runs the Const version of the method.

File: Project1/CellEdge.h (hunk 3 above)
Comment (dwatchguard) [reply Project1/CellEdge.h]: I was allowing cost to take a lambda to allow for different canPass logic, but decided that this  wasn't the right place to be making those changes.

File: Project1/MapTest.cpp [hunk 20]
```diff
@@ -15,10 +15,12 @@ bool test::runMapTest() {
...
 	UnitData ike_data = { "Ike", d, AttributeList({ 5, 0, 8, 10, 19, 4, 6, 1 }) };
+	UnitPtr mia = std::make_unique<Unit>(mia_data);
+	UnitPtr ike = std::make_unique<Unit>(ike_data);
 	auto unit_vec = std::vector<UnitData>();
 	unit_vec.push_back(mia_data);
 	unit_vec.push_back(ike_data);
-	PartyData data1 = { "Greil Mercenaries", unit_vec};
+	PartyData data1 = { "Greil Mercenaries"};
 	PartyData data2 = { "Daein" };
 	std::vector<PartyData> party_vec = std::vector<PartyData>();
```
Comment (SilensAngelusNex) [inline Project1/MapTest.cpp]: No change: just fyi you can init vector like
`std::vector<PartyData> party_vec = { data1, data2 };`

File: Project1/PartyBase.h [hunk 21]
```diff
@@ -29,7 +30,10 @@ class PartyBase {
 	iterator getPosition(const Unit& unit);
 
 	void insertUnit(UnitData unit);
-	void insertUnit(Unit& unit);
+	void insertUnit(UnitPtr& unit);
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.h]: So when you take unique_ptr by reference, it won't transfer ownership to you. The ptr is on the caller's stack, so it dies when he returns. You need to take the argument by value, which means they have to move it into your stack (since you can't copy unique_ptr), or take it by r-value reference `T&&` which makes them move it into your stack even if they _could_ copy it.

File: Project1/Mobility.h [hunk 22]
```diff
@@ -17,8 +19,8 @@ class Mobility : Component<Unit> {
 	MobilitySet& getMobilityList();
```
Comment (SilensAngelusNex) [inline Project1/Mobility.h]: `const int& getMove() const` should return by value instead of const ref

ditto for `const MobilitySet& getMobilityList() const` since they're both small

File: Project1/Mobility.cpp [hunk 23]
```diff
@@ -28,6 +30,18 @@ int& Mobility::getMove() {
...
+bool Mobility::canPass(const Unit* unit) const {
+	//TODO(Torrey): add skill Logic stuff
+	return unit == nullptr || _owner._party->hasUnit(*unit);
+}
+
+std::optional<int> Mobility::getCost(const CellEdge & edge) const {
+	std::optional<int> cost;
+	for (MobilityType type : MobilityType::list) {
+		std::optional<int> edge_cost = edge.getCost(type);
+		if (_mobility[type] && canPass(edge._cell.getTile()._unit) && edge_cost.has_value() && (cost > edge_cost || !cost.has_value())) { // if we can step on the tile
```
Comment (SilensAngelusNex) [inline Project1/Mobility.cpp]:
```
if (_mobility[type] && canPass(edge._cell.getTile()._unit)) {
    std::optional<int> edge_cost = edge.getCost(type);
    if (!cost || cost > edge_cost) {
        cost = edge_cost.value();
    }
}
````

You can swap the second if's condition with `edge_cost && (cost > edge_cost || !cost)` if you want to avoid assigning None to cost when it's already None, but that 5-condition if needs breaking up especially because you don't need edge_cost if it fails early.

**Review by SilensAngelusNex — COMMENTED** [review]
Looks good. Only important change is with the unique_ptrs in Party

File: Project1/MapTest.cpp (hunk 20 above)
Comment (dwatchguard) [reply Project1/MapTest.cpp]: I know, but when I'm testing stuff I modify the lines a bunch, and it easier when everything is on its own line.

File: Project1/PartyBase.cpp [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+	_units.emplace_back(std::make_unique<Unit>(unit));
+	_units.back()->_party = this;
+}
+
+bool PartyBase::hasUnit(const Unit& unit) const {
+	return (getPosition(unit) != cend());
+}
+
+void PartyBase::insertUnit(UnitPtr& unit) {
+	Expects(!hasUnit(*unit) && unit != nullptr);
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.cpp]: you cant have the unit since you're given an owning pointer to it

Edit: it should expect that unit->_party == nullptr, because either it was just created or another party dropped it

File: Project1/PartyBase.cpp [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+		_units.back()->_party = this;
+	}
+}
+
+UnitPtr PartyBase::dropUnit(Unit& unit) {
+	Expects(hasUnit(unit));
+	return dropUnit(getPosition(unit));
+}
+
+UnitPtr PartyBase::dropUnit(const iterator& pos) {
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.cpp]: What if they give you an iterator to a different party's unit?

File: Project1/Mobility.cpp (hunk 23 above)
Comment (dwatchguard) [reply Project1/Mobility.cpp]: canPass(edge._cell.getTile()._unit) is going to be removed anyway since Tile will be losing their unit, but I fixed this anway

File: Project1/Party.cpp [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+			unit->refresh();
+		}
+	}
+	else {
+		for (auto& unit : _units) {
+			unit->newTurn();
+		}
+	}
+}
+void Party::insertUnit(UnitPtr& unit) {
```
Comment (SilensAngelusNex) [inline Project1/Party.cpp]: You don't have to reimplement any of these. You inherited PartyBase publicly, so everyone can call public PartyBase methods on you.

File: Project1/PartyBase.h [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+	using const_iterator = std::list<UnitPtr>::const_iterator;
+
+	std::string _party_name;
+	std::list<UnitPtr> _units;
+
+	PartyBase();
+	PartyBase(std::string name);
+	PartyBase(std::string name, std::vector<UnitData> unit_data);
+	PartyBase(PartyData data);
+
```
Comment (SilensAngelusNex) [inline Project1/PartyBase.h]: You need to explicitly declare the default destructor to make it protected.
`~PartyBase() = default;`

**Review by SilensAngelusNex — CHANGES_REQUESTED** [review]
The PartyBase Party interactions need changes.

File: Project1/PartyBase.cpp (hunk 25 above)
Comment (dwatchguard) [reply Project1/PartyBase.cpp]: Iterator's don't have a way to tell ownership sadly :/

File: Project1/Party.cpp [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+}
+UnitPtr Party::dropUnit(const iterator& pos) {
+	return PartyBase::dropUnit(pos);
+}
+/*Use this function to determine when to change turns
+*/
+bool Party::isDone() const{
+	bool isDone = true;
+	for (auto& unit : _units) {
+		isDone = isDone && (unit->isTired() || unit->getHealth().isDead());
```
Comment (SilensAngelusNex) [inline Project1/Party.cpp]: You can use `&=` here

File: Project1/Party.cpp (hunk 26 above)
Comment (dwatchguard) [reply Project1/Party.cpp]: InsertUnit was protected, should I make it public?

File: Project1/PartyBase.cpp (hunk 24 above)
Comment (dwatchguard) [reply Project1/PartyBase.cpp]: Insert currently steals Units in another party, should I change that?

File: Project1/Mobility.h [hunk 29]
```diff
@@ -13,11 +13,8 @@ class Mobility : Component<Unit> {
 	Mobility(Unit& owner);
 	Mobility(Unit& owner, int move, MobilitySet mobility_type);
 
-	const MobilitySet& getMobilityList() const;
-	const int& getMove() const;
-
-	MobilitySet& getMobilityList();
-	int& getMove();
```
Comment (SilensAngelusNex) [inline Project1/Mobility.h]: Why did you remove this one? Now there's no way to change _movement.

**Review by SilensAngelusNex — APPROVED** [review]

File: Project1/Mobility.h (hunk 29 above)
Comment (dwatchguard) [reply Project1/Mobility.h]: For some reason, I thought you wanted that gone too oops. I knew I included it for a reason.
